moped-toolbox/migrate_personnel/run.py

```diff
@@ -0,0 +1,151 @@
+import argparse
+import requests
+
+from secrets import HASURA_AUTH
+
+
+def make_hasura_request(*, query, variables, env):
+    """Fetch data from hasura
+
+    Args:
+        query (str): the hasura query
+        env (str): the environment name, which will be used to acces secrets
+
+    Raises:
+        ValueError: If no data is returned
+
+    Returns:
+        dict: Hasura JSON response data
+    """
+    admin_secret = HASURA_AUTH["hasura_graphql_admin_secret"][env]
+    endpoint = HASURA_AUTH["hasura_graphql_endpoint"][env]
+    headers = {
+        "X-Hasura-Admin-Secret": admin_secret,
+        "content-type": "application/json",
+    }
+    payload = {"query": query, "variables": variables}
+    res = requests.post(endpoint, json=payload, headers=headers)
+    res.raise_for_status()
+    data = res.json()
+    try:
+        return data["data"]
+    except KeyError:
+        raise ValueError(data)
+
+
+PROJ_PERSONNEL_QUERY = """
+  query GetProjectPersonnel {
+    moped_proj_personnel(
+      where: {
+        _and: [
+          { is_deleted: { _eq: false } }
+          { _not: { moped_proj_personnel_roles: {} } }
+        ]
+      }
+    ) {
+      role_id
+      project_personnel_id
+      project_id
+      user_id
+      notes
+    }
+  }
+"""
+
+PROJ_PERSONNEL_UPDATE = """
+mutation UpdateProjectPersonnel(
+    $id: Int!,
+    $object: moped_proj_personnel_set_input!,
+    $project_roles: [moped_proj_personnel_roles_insert_input!]!,
+    $delete_ids: [Int!]!
+  ) {
+  update_moped_proj_personnel_by_pk(pk_columns: {project_personnel_id: $id}, _set: $object) {
+    project_personnel_id
+  }
+  insert_moped_proj_personnel_roles(objects: $project_roles) {
+    affected_rows
+  }
+  update_moped_proj_personnel(where: {project_personnel_id: {_in: $delete_ids}}, _set: {is_deleted: true}) {
+    affected_rows
+  }
+}
+"""
+
+
+def get_personnel_index(personnel):
+    index = {}
+    for p in personnel:
+        key = f"{p['project_id']}_{p['user_id']}"
+        index.setdefault(key, [])
```

Comment:
TIL

Reply:
was just looking for a js equivalent and i gotta say they are [pretty gross / clever](https://stackoverflow.com/questions/16107680/javascript-equivalent-of-pythons-dict-setdefault)

```javascript
d.key ??= value
```

one of those rare occasions where python has javascript beat wrt to the object API??